packages/core/src/storages/request_provider.ts

```diff
@@ -418,6 +411,11 @@ export abstract class RequestProvider implements IStorage {
             resolve(finalAddedRequests);
         });
 
+        this.inProgressRequestBatches.push(promise);
+        void promise.finally(() => {
+            this.inProgressRequestBatches = this.inProgressRequestBatches.filter((it) => it !== promise);
+        });
```

Comment:
how many items do you we expect in that array in a high concurrency run? this solution is not the best one, but if the size wont be large, we can keep it.

how is this different than a simple integer counter? that would be the most performant approach, just increment instead of push and decrement in the finally block

Reply:
The integer counter was in fact the previous implementation. However, it could not work with multiple clients, and we cannot reliably detect that - the `queueHadMultipleClients` flag is set even if the other client was a pre-migration instance of the same run, if that makes sense.

You are right that each forefront request might make us lock 25 more requests, and that could unbalance parallel instances quite a bit. Maybe we should give up "excess" requests after we're done checking for forefront requests.

Reply:
Hmm, not sure I follow why the counter wouldn't be enough, how is this better? Each client will have its own local cache (this new var). You store values in an array and wipe them based on identity, but the promises are not really used anywhere. My suggestion is doing the same, just without the memory/perf overhead.

Just to be sure, this is what I meant, it still uses the `promise.finally`:

```ts
this.inProgressRequestBatches++;
void promise.finally(() => {
    this.inProgressRequestBatches--;
});
```

Reply:
Oh damn, I'm sorry. I thought you are commenting on a different part of code - the one that handles forefront requests. If it's any help, you pushed me to tie up a loose end that I forgot about.

Regarding the batches, you're probably right :grin: